File: Lab/Lab1/code/golang/naive.go
package main

import (
	"fmt"
)

type Naive struct {
	mapping     map[int][]int
	setIDs      []int
	allValues   []int
	bound       float64
}

func NewNaive(bound float64, textSel int) *Naive {
	collector := NewCollector(textSel)
	mapping := collector.GetData()
	setIDs := collector.GetKeys()
	allValues := collector.GetAllValues()
	return &Naive{
		mapping:   mapping,
		setIDs:    setIDs,
		allValues: allValues,
		bound:     bound,
	}
}

func (n *Naive) calcSimilarity(set1, set2 map[int]bool) float64 {
	intersection := 0
	for key := range set1 {
		if set2[key] {
			intersection++
		}
	}
	union := len(set1) + len(set2) - intersection
	similarity := float64(intersection) / float64(union)
	return similarity
}

func (n *Naive) calcSimilarityAll() [][]interface{} {
	var similarityAll [][]interface{}
	for i := 0; i < len(n.setIDs)-1; i++ {
		set1 := make(map[int]bool)
		for _, item := range n.mapping[n.setIDs[i]] {
			set1[item] = true
		}
		id1 := n.setIDs[i]
		for j := i + 1; j < len(n.setIDs); j++ {
			set2 := make(map[int]bool)
			for _, item := range n.mapping[n.setIDs[j]] {
				set2[item] = true
			}
			id2 := n.setIDs[j]
			similarity := n.calcSimilarity(set1, set2)
			if similarity >= n.bound {
				similarityAll = append(similarityAll, []interface{}{id1, id2, similarity})
			}
		}
	}
	return similarityAll
}

func main() {
	naive := NewNaive(0.9, 3)
	similarityAll := naive.calcSimilarityAll()
	fmt.Println("Similarity:", similarityAll)
}

File: Lab/Lab1/code/naive.py
import sys

from collector import Collector
from tqdm import tqdm

from numba import jit


@jit(nopython=True)
def calc_similarity(set1, set2):
    """
    计算两个集合的相似度
    :param set1:
    :param set2:
    :return:
    调用关系
    calc_similarity_all
    """

    union = len(set1.union(set2))
    inter = len(set1.intersection(set2))
    similarity = float(inter / union)
    return similarity


class Naive:
    """
    用于计算naive similarity
    :param __collect: 数据收集器
    :param __bound: 相似度的阈值
    调用关系
    main.py -> write_naive -> Naive
    """

    def __init__(self, __collect, __bound=0.9):
        collector = __collect
        self.map = collector.get_map()
        self.set_ids = collector.get_keys()
        self.all_values = collector.get_all_values()
        self.bound = __bound

    def calc_similarity_all(self):
        similarity_all = []
        for i in tqdm(range(len(self.set_ids)), file=sys.stdout, desc='Calculating naive similarity'):
            set1 = set(self.map[self.set_ids[i]])
            id1 = self.set_ids[i]
            for j in range(i + 1, len(self.set_ids)):
                set2 = set(self.map[self.set_ids[j]])
                id2 = self.set_ids[j]
                similarity = calc_similarity(set1, set2)
                if similarity >= self.bound:
                    similarity_all.append(f"{id1}\t{id2}")
        return similarity_all
